### ISSUE TITLE
Add option to remove the prefixing of "friendly_name" since this affects the voice assistant

### ISSUE THREAD

**Issue by rarroyo6** [esphome/feature-requests#2476]
**Describe the problem you have/What new integration you would like**

The prefixing of "friendly_name" to entities should be optional, or have a way to disable it.
**Please describe your use case for this integration and alternatives you've tried:**

When the "friendly_name" gets prefixed to the name of an entity, especially when a ESP node handles multiple different type of entities, it can sometimes make the name incoherent. This can sometimes be acceptable to look at, but when the name is used by voice assistant, it can make the sentence completely awkward or too wordy.
I know that the name can be changed in the UI, or an alias can be created, but it should be possible to do this from the beginning.
**Additional context**

Here is an example of the problem:
I have a ESP node with a friendly name of "Pool chemical controller", then in that node I have entities "Chlorine pump", "Acid pump", plus a few others and a variety of sensors.
If I want to turn on the chlorine pump, right now I have to say "Turn on the pool chemical controller chlorine pump", I should be able to just say "Turn on the chlorine pump".

**Comment by rarroyo6** [comment]
> as a workaround, you could use the Alias feature in HomeAssistant for changing the entity name... ![grafik](https://user-images.githubusercontent.com/26391061/285405564-f317a991-36ce-43b3-b381-23b419f50258.png)

I mentioned in the original post, that I know that I could add an alias, but the idea is to make it simpler, and not to have to take extra steps in order to make it work right.

**Comment by clinta** [comment]
The purpose is incredibly useful to me. I have several of the same device type, like an s31 relay for example, and I use templates to mange the configuration. So in my template I can have something simple like `name: Button`, then if the friendly name is "Living Room Lamp", in home assistant I get "switch.living_room_lamp_button".

I depend on this, and if it is optional it should be opt-out, except in one case. In the case of an empty name like `name: None` or `name: ""`, the documented behavior is that this should use the friendly_name, but because of the prefixing it is repeated. This should be fixed by default. I should not end up with an entity of `switch.living_room_lamp_living_room_lamp`, when I use `name: None` for the relay. It should just be `switch.living_room_lamp`.

**Comment by Linwood-F** [comment]
First, let me say I hate the whole concept of the automatically prefixed friendly name, and it was absolutely the most disruptive rollout of anything I have encountered in HA in many years of using it. I think many of us use various tricks to work around it (e.g. using name with a single space as opposed to empty strings).

That said, I think some care has to be taken in changing it yet again. The other thing that has happened is 3rd parties have rolled out very useful (and widely used) git based implementations, which make it more complicated to change THEIR code (e.g. forking and merging updates -- and HA thinks YAML is hard to edit!). 

For example, I personally would like to see any such control over whether or not to prefix be locally controlled, and by that I mean in different parts of the same esphome program.  If I have includes from GIT, and local includes of my own standard code in the same file, being able to have some kind of pragma declaration might work, where one is #PREFIX OFF and #PREFIX ON (clearly not that syntax, just as an example) that has a respected scope. 

But please... stop breaking existing code for this.  It was a royal mess most have gotten past, I am sure it was well intentioned but it was a mess.  Don't make another with trying to add more lipstick to this pig.

**Comment by rarroyo6** [comment]
This change has affected Voice Assistant, Automations, Scripts, and Dashboards.
I'm still finding some rarely used automations that when I realize they didn't work and check why, it turns out that they have the old name before it was prefixed.
I've also come across some old EspHome nodes that had not been updated, and were working fine. After updating them, the names changed and created more problems.

**Comment by alexruffell** [comment]
I have this issue too. My most common use case is when I create devices in ESPHome that can be turned on or off, or opened and closed.

I am modifying the default YAML of the ratgdo and since my ratgo is called "Overhead Garage Door", I do not want anything added to the entity name... in other words it must match the device name. I have tried a few things:

name: None (does not work, I just get "none" as a name)
name: "" (does not work)
name: " " (works as desired but it seems more like exploiting a bug)

My end goal is for the cover entity that controls the "Overhead Garage Door" to be called `cover.overhead_garage_door` which is also the name of the device (ratgdo). Simply put, some way for an entity to get the device name and nothing else would be great. Before the automated prepending was done, I had full control and it worked perfectly... sadly the "facilitation" works in most cases but caused major annoyance in these particular cases (frequent in my case).

Summary: What @clinta said sounds perfect - https://github.com/esphome/feature-requests/issues/2476#issuecomment-1844034117

**Comment by kooswillem** [comment]
You can overwrite (change) the entity ID in 'Developer Tools -> States page' by finding the entity, clicking on the 'i' icon (more information) and then in the pop-up window click the sprocket icon. Change the entity ID and confirm by clicking 'Update' in the left bottom corner of this pop-up window.

**Comment by nagyrobi** [comment]
> You can overwrite (change) the entity ID in 'Developer Tools -> States page' by finding the entity, clicking on the 'i' icon (more information) and then in the pop-up window click the sprocket icon. Change the entity ID and confirm by clicking 'Update' in the left bottom corner of this pop-up window.

You should avoid doing that on long term and with many entities. If for any reason in the future you need to remove and re-add the ESPHome node to HA, you'll have to do the renames again manually, for all your automations and data history to be functional again.

**Comment by Unfocused** [comment]
@alexruffell wrote:
> name: None (does not work, I just get "none" as a name)

This should work - I'd suggest filing a separate bug report (not feature request) just for this.

**Comment by Unfocused** [comment]
With specifying `None` as the name, some of the use-cases mentioned here would work well with the sub-device/multi-device config being added in this PR:
- https://github.com/esphome/esphome/pull/6484

For background discussion on that, see:
- https://github.com/esphome/feature-requests/issues/1335

**Comment by DrBlokmeister** [comment]
I'm also really surprised that this change was suddenly rolled out and I find that all my ESPHome sensors now are named something like sensor.plug_nas_plug_nas_power, instead of sensor.plug_nas_power and I would need to manually modify around 20 sensors per device. 

Please give us the option to remove prefixing of friendly_name!

**Comment by s-wachspress** [comment]
I found this thread while looking for a solution to this issue and discovered that if you use a space for the friendly name, the whitespace prefix will get trimmed automatically, avoiding any automated prefix being added to the other entity's name properties:
```
esphome:
  friendly_name: " "

button:
  - platform: template
    name: "TV Power"
    id: tv_power_on
```

Automatically shows up with the name "TV Power" in my list of Entities.
Notably, this won't work if you provide a completely empty string: ""

**Comment by rarroyo6** [comment]
> I found this thread while looking for a solution to this issue and discovered that if you use a space for the friendly name, it get's recognized as a valid prefix, but the whitespace gets trimmed from the entity name automatically,
> 
> ```
> esphome:
>   friendly_name: " "
> 
> button:
>   - platform: template
>     name: "TV Power On"
>     id: tv_power_on
> ```
> 
> This config automatically generated an entity with the name "TV Power On" in my list of Entities.
> 
> Notably, this won't work if you provide a completely empty string: "" I'm pretty sure you also need to Clean Build Files before applying this config change.

The blank space works for the individual entity, but it defeats the purpose of the friendly name, which is to name the node.
You could have a node with a generic friendly name, to identify it in the ESPHome dashboard, but that node can have multiple unrelated entities, which won't make any sense when the friendly name gets appended.
For example: I have a node with the friendly name "Backyard Sonoff 4ch relay board", which makes it easy to identify on the ESPHome dashboard, but this node controls multiple unrelated entities, like Fence lights, Tree lights, Fairy lights, and Pond light, when the friendly name gets appended, the names become "Backyard Sonoff 4ch relay board Fence lights", etc...
This should be an option, but it should not be forced.
I have 62 nodes, and more planned, and most have multiple entities. It becomes a hassle to manually rename the entities after making a change, or after removing and adding it again.

**Comment by clinta** [comment]
@rarroyo6 did you intend to close this?

**Comment by DrBlokmeister** [comment]
@rarroyo6 please re-open the issue as it has not been solved yet AFAIK.

**Comment by rarroyo6** [comment]
Sorry about that, I accidentally clicked on the submit and close button.

**Comment by Naguz** [comment]
Is this on any roadmap? I have had to stop using friendly name altogether. This gives a slightly less appealing, but at least sensible entity names. For any device with more than one entity I find this to be an absolute necessity, as the hack with " " as entity name obviously does not work for these.

**Comment by FreAcX** [comment]
Vote up

**Comment by sumasage** [comment]
vote up +1

**Comment by mainakae** [comment]
Vote up +1

**Comment by SparkyDan555** [comment]
I agree the friendly name shouldn't be added to entity names. It is supposed to prevent _"less intuitive names and a less polished experience in Home Assistant"_ but in reality it makes things more messy:

<img width="508" alt="Image" src="https://github.com/user-attachments/assets/e206bc0a-42b6-4592-a5d1-5dd1b09964a6" />

In Apple Home:
<img width="159" alt="Image" src="https://github.com/user-attachments/assets/0dda4822-28f7-4a94-81fb-f184c0eb6651" />

**Comment by juicejuice** [comment]
I have a similar issue to the above. I have an ESPHome device like "Hallway Switch". It controls four different lights that are in totally different parts of the house. Hallway Switch Side Gate Light doesn't really roll off the tongue...

**Comment by nagyrobi** [comment]
Same here.

**Comment by Makr91** [comment]
This seems to have become a bigger problem starting in April this year

https://github.com/esphome/feature-requests/issues/3174
https://github.com/esphome/feature-requests/issues/2732

**Comment by caco3** [comment]
Oh bugger, it even gets odder because it is not consistent.
I just started using some smart plugs driven by ESPHome. Points out the shown duplication depends on the used view:
### Masonry View
Entity card:

<img width="503" height="88" alt="Image" src="https://github.com/user-attachments/assets/68c27b2e-d5bc-4d72-9d43-662e5801d232" />

### Sections View
Entity card:

<img width="391" height="85" alt="Image" src="https://github.com/user-attachments/assets/1f7a6a2a-4fce-4db3-b340-23755086c9bc" />

---

I am using
-  Core 2025.7.1
- Frontend 20250702.1

**Comment by anonymousaga** [comment]
This issue can be closed.

We now have [sub-devices](https://esphome.io/components/esphome/#sub-devices), allowing for each entity to have its own device. Set the sub-device friendly name to the entity name you want, and set the entity friendly name to `None`.

The issue of multiple entities receiving the same prefix is now solved with this in a clean way. Ive migrated all my devices to this system with no issues.

**Comment by DrBlokmeister** [comment]
I disagree.

The issue is that the friendly name is always prefixed to devices, regardless of whether we want this or not. I think that we should have the freedom to choose our entity name completely.

**Comment by MarkProminic** [comment]
I agree with @DrBlokmeister

**Comment by Linwood-F** [comment]
I think a lot of people agree, this was a massively disruptive rollout (and seemed to also affect mqtt), but the powers that be made it pretty clear "This is a good idea and this is how it will be" and were not receptive then to feedback -- after years of it cooking, I doubt seriously they will agree to change it.

I just ran across it again, built something and ended up with something like Network Closet Fan Fan and had to go back and redo the names.  It's just plain stupid we are writing the YAML directly and cannot actually name things the way we want, the software knows better.

**Comment by rarroyo6** [comment]
> This issue can be closed.
> 
> We now have [sub-devices](https://esphome.io/components/esphome/#sub-devices), allowing for each entity to have its own device. Set the sub-device friendly name to the entity name you want, and set the entity friendly name to `None`.
> 
> The issue of multiple entities receiving the same prefix is now solved with this in a clean way. Ive migrated all my devices to this system with no issues.

This doesn't solve the issue.
I would still like to use the `friendly_name` to organize the ESP node in the dashboard, and have a separate name for each of the devices it controls.
If we use `None` for the friendly name, it defeats the purpose.
The cleanest way would be to default to the way it is now BUT allow us to opt-out of having the `friendly_name` automatically appended to the device name!
**Every time a add a new ESP node, I have to go through the hassle of having to rename every single entity.**
This cannot be that difficult to do!!!

**Comment by DrBlokmeister** [comment]
I have been brainstorming a bit about this issue. I don't think this can easily be fully fixed in the Home Assistant integration. It is possible to fully enable or disable this option by default, by making a hard change in the code. So we would basically roll back prefixes. I don't think this is an option.
The second option is to change the ESPHome integration without touching ESPHome itself. We could then set a per-device configuration option to enable/disable prefixing. However, this can only be done after the device has been created, which means that the entity ids have been created. So this doesn't fix the issue.
The last option is to add an option in the ESPHome config itself. For example we could add this:
```yaml
esphome:
  name: livingroom
  friendly_name: Living Room Controller
  ha_prefix_entities: false   # new key, default true
```
But this would require a change in both ESPHome as in the Home Assistant integration of ESPHome.

I asked ChatGPT to make a short summary and a more detailed requirements list. If possible, I would like some comments on this, as I'm not too familiar with neat code development.

* Add `ha_prefix_entities: bool` to ESPHome YAML and propagate it as an **optional** field in `DeviceInfo` over protobuf.
* Update **aioesphomeapi** to expose the field (nullable).
* Update **Home Assistant (esphome)** to read that field at discovery and seed the per-entry **Options** flag that controls prefixing (already implemented in your `entity.py`).
* Use **optional proto field** + **defaults** so **old/new** versions interoperate with no errors.
* New devices get the right behavior *from the start*; existing devices keep stable IDs, with the option available to adjust display names and control future IDs.

<details>
<summary>**Click here for the full suggested change list. Warning: long!**</summary>

# What we’re adding

A new top-level ESPHome config option, e.g.

```yaml
esphome:
  name: livingroom
  friendly_name: Living Room Controller
  ha_prefix_entities: false   # new (default true)
```

This is **device-authored intent** that Home Assistant (HA) should respect on first import and expose in Options thereafter.

---

# The moving parts (and how to keep them backward/forward compatible)

## 1) ESPHome firmware (Dashboard + codegen)

**Goal:** accept `ha_prefix_entities` in YAML and include that preference in the node’s **DeviceInfo** metadata sent over the API.

* **YAML schema:** Add a boolean key `ha_prefix_entities` with default `true`.
* **Runtime storage:** Carry this value in the internal config object used to fill out API **DeviceInfo**.
* **No behavior change on device:** This is purely metadata for HA; ESPHome itself doesn’t rename anything.

### Back/forward-compat on the device

* Devices can safely **send** extra fields in protobuf; old clients just ignore them (classic proto wire-compat).

## 2) ESPHome API (protobuf)

**Goal:** convey the preference across the wire without breaking old clients.

* **File:** `api.proto`

* **Message:** extend `DeviceInfo` (or a similar always-sent info message) with an **optional bool** field:

  ```proto
  message DeviceInfo {
    // existing fields...
    optional bool ha_prefix_entities = 42; // pick an unused high tag number
  }
  ```

  * **Default semantics:** if the field is **absent**, it means “unknown → assume TRUE”.
  * Using a **new tag number** + `optional` keeps proto backward/forward compatible:

    * **Old clients** (old aioesphomeapi/HA) don’t know the field → silently ignore.
    * **New clients** see it when the device is updated; when talking to old devices, they just won’t see the field.

* **Version gating (optional but nice):**

  * Add a **capabilities bit** (e.g., `capabilities2.has_ha_prefix_entities`) or bump the ESPHome API version string.
  * Not required for safety (proto already safe), but helps telemetry/debugging.

## 3) aioesphomeapi (Python client)

**Goal:** expose the new field to HA; stay safe if device or HA is old.

* Regenerate client stubs (or add a nullable attribute) so `DeviceInfo.ha_prefix_entities` is available (type: `Optional[bool]`).
* Keep **default None** if the field is not present (older firmware).
* No behavioral change otherwise.

## 4) Home Assistant integration (esphome)

**Goal:** read the preference at **first discovery** and use it as the initial value for the per-entry **Options** flag you added earlier.

* **config_flow.py** (discovery/user flow):

  * After fetching `device_info = await cli.device_info()`, read:

    ```python
    device_pref = getattr(device_info, "ha_prefix_entities", None)
    initial_prefix = DEFAULT_PREFIX_ENTITY_BY_DEVICE if device_pref is None else bool(device_pref)
    ```
  * When creating the ConfigEntry:

    ```python
    options = {
      CONF_ALLOW_SERVICE_CALLS: DEFAULT_NEW_CONFIG_ALLOW_ALLOW_SERVICE_CALLS,
      CONF_PREFIX_ENTITY_BY_DEVICE: initial_prefix,
    }
    ```
  * If user later changes it in HA Options, HA wins (Options override device hint).

* **entity.py** (already done): uses the option to set `_attr_has_entity_name` and affect **new** entity_id generation.

### Back/forward-compat in HA

* **New HA + old ESPHome:** `device_info.ha_prefix_entities` is absent → default to **True** (current behavior).
* **Old HA + new ESPHome:** Proto field is ignored by the old client; device still works (no crash).
* **Mixed upgrades:** All safe; field is optional and HA defaults preserve current behavior.

---

# UX / migration behavior

* **First import (brand-new device):**

  * HA reads the device’s preference and seeds the ConfigEntry option accordingly.
  * Entities are created with/without the prefix **from the start** (no rename pain).
* **Existing devices already added to HA:**

  * Nothing auto-renames (HA policy). The new field simply **pre-populates** HA Options for future re-adds.
  * If you want to align legacy entity IDs, you’ll do a one-time **bulk rename** (scriptable). Display names change immediately when you toggle the option.

---

# Edge cases & considerations

* **Multiple sub-devices:** Your `entity.py` gates `_attr_has_entity_name` after resolving `device_name` per sub-device — still correct. The device-wide preference applies uniformly.
* **Dashboards / voice:** Disabling prefixing affects composed display names. Voice assistants that rely on friendly names will see the shorter names (often desirable).
* **Conflicts:** Without prefixing, duplicate entity names across devices can create suggested-ID collisions. HA handles this by deduping (e.g., suffixing `_2`), but it’s worth documenting.

---

# Testing plan (recommended)

1. **ESPHome unit:**

   * YAML parse → `ha_prefix_entities` stored; proto DeviceInfo includes the field correctly for true/false; absent when not set (default true).
2. **aioesphomeapi:**

   * Decode both messages with and without field; ensure Python returns `None` vs `True/False`.
3. **HA integration:**

   * Simulate DeviceInfo with `ha_prefix_entities=None/True/False`; validate ConfigEntry options seeded accordingly; assert resulting entity friendly names and suggested object IDs.

---

# Rollout strategy

1. Merge **ESPHome (proto change + device filling)** first; publish.
2. Merge **aioesphomeapi** with regenerated stubs; HA will pin a min version in requirements.
3. Merge **HA** reading the field to seed options.
   All permutations still run safely due to proto optionality + HA defaults.

</details>

**Comment by rarroyo6** [comment]
This is what I suggested originally, just add a flag in the YAML, default it to TRUE so nothing changes unless you want to disable it manually, then you set it to FALSE.

**Comment by MarkProminic** [comment]
They do not think this is a big enough of an issue to give this any attention, they basically dropped a hand grenade and ran for the hills.

**Comment by MarkProminic** [comment]
I give up, now its prefixing the Area too, i guess I will just never use the Voice Assistant Previews I bought, without having to say a incredibly long name with duplicative words. This whole Friendly Name thing just plain sucks. I miss the days of yaml.

**Comment by DrBlokmeister** [comment]
This is really getting crazy. Now my sensor is called sensor.baby_room_aqm_baby_room_baby_room_aqm_absolute_humidity.

How can we get this noticed by the devs?

**Comment by randybb** [comment]
Because HA devs are living in their bubble and have no clue about esphome. I get it, in most cases we have area in the device name, but I feel it stating to be crazy. Basically you have a device with name "Office Light", then the area will be "Office" and friendly name 'Light". In such case it is fine. But what if you have a device that you are moving between areas? You can switch area and done (and the auto dashboard will be updated automagically), but with prefixed area will the entity id carring the original area, which is wrong.
And if you are using that autofeature and have devices that are monitoring/controlling multiple areas, then you need to redo everything with subdevices, otherwise it will be mess.
I really miss the old times when the entity id was fully in our hands and when we wanted to consolidate devices we could move that entity from one device to another without changing its entity id and loosing history.

**Comment by DrBlokmeister** [comment]
Yeah it is also named here: https://github.com/OpenHomeFoundation/roadmap/issues/73. I hope they roll it back.
